appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,730 @@
+ * game. Should be less than MAX_POPULATION * number of players.
+ * @type {number}
+ */
+Genetics.Cage.START_MICE_PER_PLAYER = 2;
+
+/**
+ * Mapping of player id to and array with player name and code.
+ * @type {!Object.<number, Array.<string>>}
+ */
+Genetics.Cage.PLAYERS = {};
+
+/**
+ * Whether the simulation is running without a visualization.
+ * @type {boolean}
+ */
+Genetics.Cage.HEADLESS = false;
+
+/**
+ * List of mice currently alive.
+ * @type {!Array.<Genetics.Mouse>}
+ * @private
+ */
+Genetics.Cage.aliveMice_ = [];
+
+/**
+ * Mapping of mouse id to mouse for quick lookup.
+ * @type {!Object.<number, Genetics.Mouse>}
+ * @private
+ */
+Genetics.Cage.miceMap_ = {};
+
+/**
+ * Mapping of mouse id to PID of life simulation for that mouse queued to run.
+ * @type {!Object.<number, number>}
+ * @private
+ */
+Genetics.Cage.mouseLife_ = {};
+
+/**
+ * List of events to be visualized.
+ * @type {!Array.<!Event>}
+ */
+Genetics.Cage.EVENTS = [];
+
+/**
+ * Template for event properties.
+ * @type {Object.<string, Array.<string>>}
+ */
+Genetics.Cage.EVENT_TEMPLATE = {
+  'ADD': ['TYPE', 'MOUSE', 'PLAYER_NAME'],
+  'START_GAME': [],
+  'FIGHT': ['TYPE', 'ID', 'RESULT', 'OPT_OPPONENT'],
+  'MATE': ['TYPE', 'ID', 'RESULT', 'OPT_PARTNER', 'OPT_OFFSPRING'],
+  // A mouse dies after using all fight and mate opportunities.
+  'RETIRE': ['TYPE', 'ID'],
+  // The oldest mouse dies when there is no room for it.
+  'OVERPOPULATION': ['TYPE', 'ID'],
+  // A mouse dies if it throws an exception.
+  'EXPLODE': ['TYPE', 'ID', 'SOURCE', 'OPT_CAUSE'],
+  // A mouse dies if it does not complete execution.
+  'SPIN': ['TYPE', 'ID'],
+  'END_GAME': ['TYPE', 'CAUSE', 'PICK_FIGHT_WINNER', 'CHOOSE_MATE_WINNER',
+    'MATE_ANSWER_WINNER']
+};
+
+/**
+ * Creates an event.
+ * @param {string } type The type of event.
+ * @param {...string|number|Genetics.Mouse} var_args The properties on
+ * the event.
+ * @constructor
+ */
+Genetics.Cage.Event = function(type, var_args) {
+  var template = Genetics.Cage.EVENT_TEMPLATE[type];
+  for (var i = 0; i < template.length; i++) {
+    this[template[i]] = arguments[i];
+  }
+};
+
+/**
+ * Adds the event to the queue depending on the mode the game is running as.
+ */
+Genetics.Cage.Event.prototype.addToQueue = function() {
+  // Headless does not need to keep track of events that are not end game.
+  if (Genetics.Cage.HEADLESS &&
+    this.TYPE != Genetics.Cage.EVENT_TEMPLATE.END_GAME) {
+    return;
+  }
+  Genetics.Cage.EVENTS.push(this);
+};
+
+/**
+ * Time to end the battle.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.endTime_ = 0;
+
+/**
+ * Time limit of game (in milliseconds).
+ * @type {number}
+ */
+Genetics.Cage.GAME_TIME_LIMIT = 5 * 60 * 1000;
+
+/**
+ * Time limit for mouse function execution in number of interpreter steps (100k
+ * ticks runs for about 3 minutes).
+ * @type {number}
+ */
+Genetics.Cage.FUNCTION_TIMEOUT = 100000;
```

Comment:
Optional: FUNCTION_TIMEOUT_NUM_STEPS. Not strictly necessary, I think, but it seems useful to state the "unit".